storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java

```diff
@@ -0,0 +1,222 @@
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] strSplit = str.split(" ");
+                if (!strSplit[2].equals("cgroup")) {
+                    continue;
+                }
+                String name = strSplit[0];
+                String type = strSplit[3];
+                String dir = strSplit[1];
+                Hierarchy h = new Hierarchy(name, CgroupUtils.getSubSystemsFromString(type), dir);
+                hierarchies.put(type, h);
+            }
+            return new ArrayList<Hierarchy>(hierarchies.values());
+        } catch (Exception e) {
+            LOG.error("Get hierarchies error {}", e);
+        }
+        return null;
+    }
+
+    @Override
+    public Set<SubSystem> getSubSystems() {
+
+        Set<SubSystem> subSystems = new HashSet<SubSystem>();
+
+        try (FileReader reader = new FileReader(CgroupUtils.CGROUP_STATUS_FILE);
+             BufferedReader br = new BufferedReader(reader)){
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] split = str.split("\t");
+                SubSystemType type = SubSystemType.getSubSystem(split[0]);
+                if (type == null) {
+                    continue;
+                }
+                subSystems.add(new SubSystem(type, Integer.valueOf(split[1]), Integer.valueOf(split[2]),
+                        Integer.valueOf(split[3]).intValue() == 1 ? true : false));
+            }
+            return subSystems;
+        } catch (Exception e) {
+            LOG.error("Get subSystems error {}", e);
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isSubSystemEnabled(SubSystemType subSystemType) {
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
+            if (subSystem.getType() == subSystemType) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Hierarchy getHierarchyWithSubSystem(SubSystemType subSystem) {
+        return getHierarchyWithSubSystems(Arrays.asList(subSystem));
+    }
+
+    @Override
+    public Hierarchy getHierarchyWithSubSystems(List<SubSystemType> subSystems) {
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        for (Hierarchy hierarchy : hierarchies) {
+            Hierarchy ret = hierarchy;
+            for (SubSystemType subSystem : subSystems) {
+                if (!hierarchy.getSubSystems().contains(subSystem)) {
+                    ret = null;
+                    break;
+                }
+            }
+            if (ret != null) {
+                return ret;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isMounted(Hierarchy hierarchy) {
+
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        if (Utils.CheckDirExists(hierarchy.getDir())) {
```

Comment:
Also, getHierarchies is going and reading from the file system. If getHierarchies is being called too often, you should consider caching.